### ISSUE TITLE
GPU OOM for running GCN example code

### ISSUE THREAD

**Issue by sungyongs** [dmlc/dgl#997]
## 🐛 Bug

I tried to follow GCN code described in https://docs.dgl.ai/en/latest/tutorials/models/1_gnn/1_gcn.html#sphx-glr-tutorials-models-1-gnn-1-gcn-py

When I run the training snippet after adding one line, OOM occurs.

```python
import time
import numpy as np
g, features, labels, mask = load_cora_data()
optimizer = th.optim.Adam(net.parameters(), lr=1e-3)
dur = []
for epoch in range(30):
    if epoch >=3:
        t0 = time.time()

    logits = net(g, features)
    logp = F.log_softmax(logits, 1)
    loss = F.nll_loss(logp[mask], labels[mask])

    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    if epoch >=3:
        dur.append(time.time() - t0)

    print("Epoch {:05d} | Loss {:.4f} | Time(s) {:.4f}".format(
            epoch, loss.item(), np.mean(dur)))

    logits = net(g, features)    #### Added one line

```
## To Reproduce

1. Add `logits = net(g, features)` in the training loop.
1. Run the training snippet.
1. 

## Expected behavior
`logits = net(g, features)` is just a forward computation and it should not cause OOM.
I guess that DGL internally reallocates memory for the same parameters.
This one line is useful to check `validation error` which is used to know when overfitting happens.

## Environment

 - DGL Version (e.g., 1.0): 0.4
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3): PyTorch 1.2.0
 - OS (e.g., Linux): Linux
 - How you installed DGL (`conda`, `pip`, source): `conda`
 - Build command you used (if compiling from source): 
 - Python version: 3.6.9
 - CUDA/cuDNN version (if applicable): CUDA 9.2
 - GPU models and configuration (e.g. V100): GTX 1080 Ti
 - Any other relevant information:

**Comment by sungyongs** [comment]
In fact, I found that 

```python
for epoch in range(30):

    logits = net(g, features)
```
Running above snippet causes OOM which is not expected.
However, when I put `loss.backward()` after computing `loss`, it is working well.
So, I guess `backward()` did something (removing unused parameters?).

Anyone knows how to handle this issue?

**Comment by VoVAllen** [comment]
Hi,

Sorry for the inconvenience. This is mainly because we stored some cache in forward function. Then if you didn't call backward function, the cache will keep increasing. PyTorch could handle this case well but we have extra non-Tensor object stored for cache, and they are not properly cleared now. We are still trying to figure out the solution.

For now, if you don't need backward function, you could add `with torch.no_grad()` at loop's beginning to solve this issue.
